### PR TITLE
Download WDL file from method detail page [GAWB-239]

### PR DESCRIPTION
**Pull request by aednichols**
This implementation is one of the strategies suggested in [GAWB-2974](https://broadinstitute.atlassian.net/browse/GAWB-2974).

Needs PO review.

Looks like this:

screen shot 2018-01-26 at 5 22 04 pm

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [x] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by zarsky-broad** [comment]
Would something like [this](https://jsfiddle.net/UselessCode/qm5AG/) be simpler?

**Comment by aednichols** [comment]
Seems like the same thing, except the download doesn't start automatically when you click the button?

**Comment by zarsky-broad** [comment]
I mean, you'd create the file and initiate the download both on click. I just mean to say we have the wdl in memory already, why not just package that and download it, rather that re-authing the whole thing? That is to say, we'd be essentially cut and pasting it into a file for them, rather than going through the whole dance under the hood, which will be harder to maintain.

**Comment by aednichols** [comment]
> rather that re-authing the whole thing

Not sure what you’re referring to. Let’s chat Monday.

**Comment by davidangb** [comment]
code looks good, works well for me locally. I have two usability suggestions but defer to our UX experts:
1. I'd like the filename to include the snapshotId; I think this is important for methods that have multiple snapshots.
2. I really want to see the download link on the WDL tab - either instead of or in addition to where it is currently, on the Summary tab.

Curious to hear the outcome of the discussion about Isaac's suggestion; I'm also not seeing a huge difference between the two implementations.

**Comment by aednichols** [comment]
@davidangb added a link on the WDL tab, added snapshot to filename (maybe we want namespace as well?). 

Public:
screen shot 2018-01-29 at 1 40 40 pm

Non-public:
screen shot 2018-01-29 at 1 40 56 pm

**Comment by zarsky-broad** [comment]
Seems that I was somewhat confused, LGTM.

**Review by davidangb — APPROVED** [review]
thanks for the changes, :+1: pending PO review!

**Comment by aednichols** [comment]
@davidangb @zarsky-broad - I worked with Moran to fix up the styling, also dropped the sidebar button. Substantially all of the code has changed, so definitely need re-review on this one.

File: src/cljs/main/broadfcui/page/method_repo/method/details.cljs
Comment (zarsky-broad) [inline src/cljs/main/broadfcui/page/method_repo/method/details.cljs]:
```cljs
{:margin "1.5rem 0 0.5rem 1.5rem"}
```

File: src/cljs/main/broadfcui/common/icons.cljs
Comment (zarsky-broad) [inline src/cljs/main/broadfcui/common/icons.cljs:63]: Should this be a `defn-`? Not a leading question, just asking.

File: src/cljs/main/broadfcui/common/links.cljs
Comment (zarsky-broad) [inline src/cljs/main/broadfcui/common/links.cljs]: If you change the signature to `[label url & [filename]]` this will work for external links, too.

File: src/cljs/main/broadfcui/common/links.cljs
Comment (zarsky-broad) [inline src/cljs/main/broadfcui/common/links.cljs]: I prefer to invoke this as `(js/URL.createObjectURL payload-blob)`.

**Review by zarsky-broad — APPROVED** [review]

File: src/cljs/main/broadfcui/common/icons.cljs
Comment (aednichols) [reply src/cljs/main/broadfcui/common/icons.cljs:63]: I figured people might want to use it for one-off icons that aren't worth creating a function for

File: src/cljs/main/broadfcui/common/links.cljs
Comment (aednichols) [reply src/cljs/main/broadfcui/common/links.cljs]: By external do you mean... files hosted outside of FireCloud? Why would `filename` be optional in that case?

Or do you want to create a new function called "create any link you want" that takes all the possible attributes, and replaces the bodies of `create-external` and `create-download`?

File: src/cljs/main/broadfcui/common/links.cljs
Comment (aednichols) [reply src/cljs/main/broadfcui/common/links.cljs]: Gets rid of the warning, so I'm for it. Thanks.

File: src/cljs/main/broadfcui/page/method_repo/method/details.cljs
Comment (aednichols) [reply src/cljs/main/broadfcui/page/method_repo/method/details.cljs]: Yup

File: src/cljs/main/broadfcui/common/links.cljs
Comment (aednichols) [reply src/cljs/main/broadfcui/common/links.cljs]: discussed in person, implemented by making the function multi-arity

File: src/cljs/main/broadfcui/page/method_repo/method/details.cljs
Comment (davidangb) [inline src/cljs/main/broadfcui/page/method_repo/method/details.cljs:123]: it would be nice to revoke the object url when navigating away from the WDL tab ... but that looks kinda awkward to do. I wouldn't hold this PR for that.

**Review by davidangb — APPROVED** [review]
👍

File: src/cljs/main/broadfcui/page/method_repo/method/details.cljs
Comment (zarsky-broad) [reply src/cljs/main/broadfcui/page/method_repo/method/details.cljs:123]: Yeah, I'd think the way to do that would be to make `create-download-from-object` return a Component, so that we could add a lifecycle hook. Might not be a bad idea, tbh.

Comment (aednichols) [reply src/cljs/main/broadfcui/page/method_repo/method/details.cljs:123]: This was an excellent idea. Here's how React behaves when switching snapshots:
```
"cleaning up blob:http://local.broadinstitute.org/6b3e7509-a798-449e-8ed7-de892e723f17"
"creating url blob:http://local.broadinstitute.org/57201921-7f2b-4ad1-b170-37042255b80e"
"cleaning up blob:http://local.broadinstitute.org/57201921-7f2b-4ad1-b170-37042255b80e"
"creating url blob:http://local.broadinstitute.org/8da51905-251f-4549-8bb7-10b428ad6a35"
"cleaning up blob:http://local.broadinstitute.org/8da51905-251f-4549-8bb7-10b428ad6a35"
"creating url blob:http://local.broadinstitute.org/79e21db9-5857-4008-bf38-7f32a14f3467"
"cleaning up blob:http://local.broadinstitute.org/79e21db9-5857-4008-bf38-7f32a14f3467"
"creating url blob:http://local.broadinstitute.org/a225472a-b31b-4410-ac7c-afd3d3b1c5d6"
"cleaning up blob:http://local.broadinstitute.org/a225472a-b31b-4410-ac7c-afd3d3b1c5d6"
"creating url blob:http://local.broadinstitute.org/a5c090de-cd29-4263-a336-109f888da2a5"
"cleaning up blob:http://local.broadinstitute.org/a5c090de-cd29-4263-a336-109f888da2a5"
"creating url blob:http://local.broadinstitute.org/87467c6c-03e9-4b97-90e2-b2839e90390e"
```
Without this tweak, we would be leaking 6 x (some potentially-large WDL) into memory!

**Review by davidangb — APPROVED** [review]
:+1: nice job on the revoke! (did not test)

**Comment by aednichols** [comment]
jenkins retest

**Comment by aednichols** [comment]
Failed at `OrchestrationApiSpec`

https://fc-jenkins.dsp-techops.broadinstitute.org/job/firecloud-ui-fiab-test-runner/167/consoleFull

**Comment by aednichols** [comment]
https://fc-jenkins.dsp-techops.broadinstitute.org/job/firecloud-ui-fiab-test-runner/216/consoleFull

Nicole says OK to go

**Pull request merged**
